tests/integration/ha_tests/test_ha.py

```diff
@@ -586,3 +588,48 @@ async def test_network_cut(ops_test: OpsTest, continuous_writes, chaos_mesh):
 
     # verify that old primary is up to date.
     await verify_writes(ops_test)
+
+
+async def test_storage_re_use(ops_test, continuous_writes):
+    """Verifies that database units with attached storage correctly repurpose storage.
+
+    It is not enough to verify that Juju attaches the storage. Hence test checks that the mongod
+    properly uses the storage that was provided. (ie. doesn't just re-sync everything from
+    primary, but instead computes a diff between current storage and primary storage.)
+    """
+    app = await get_application_name(ops_test, APP_NAME)
+
+    # removing the only replica can be disastrous
+    if len(ops_test.model.applications[app].units) < 2:
+        await ops_test.model.applications[app].add_unit(count=1)
+        await ops_test.model.wait_for_idle(apps=[app], status="active", timeout=1000)
+
+    # remove a unit and attach it's storage to a new unit
+    current_number_units = len(ops_test.model.applications[app].units)
+    await scale_application(ops_test, app, current_number_units - 1)
+    await ops_test.model.wait_for_idle(
+        apps=[app], status="active", timeout=1000, wait_for_exact_units=(current_number_units - 1)
+    )
+
+    # k8s will automatically use the old storage from the storage pool
+    removal_time = datetime.now(timezone.utc).timestamp()
+    await scale_application(ops_test, app, current_number_units)
+    await ops_test.model.wait_for_idle(
+        apps=[app], status="active", timeout=1000, wait_for_exact_units=(current_number_units)
+    )
+
+    new_unit = get_newest_unit(ops_test, app)
```

Comment:
**suggstion** maybe we could do a further assertion here: we could get the new_unit before scaling down, and after scaling up, and assert that the unit has the same ordinal number. I'm not sure there is a property to check that they have a different hash or something (since they are two different pods...), but if there is, I would check for this

Reply:
I like this suggestion!  But for the purpose of this test, I believe this function is sufficient and would like to keep it simple. (i.e. we want to get the most recently added unit from juju, if we were to scale the app by more than one unit then I think there might be some inconsistencies with which unit is newer.) I'll add a comment to the code to make this more clear. Thanks again Enrico :)